# Utility Maximization and Income Constraint

The utility function in economics is a representation of the individual preferences of goods or services apart from the explicity monetary value of those goods or services. This utility function measure how much someone desires some good or service. This concept of utility helps to explain and measure mathematically different social constructs. 

The utility maximization consists on obtaining the values of units of two or more goods or services that maximize the utility of a certain individual. But this function needs to be constraint, because in real life all the purchases of goods and services are constrained by the budget or income that a person has. The constraint is that the income ($I$) needs to be equal  to the money spent on product, which is equal to the number of units of that product multiplied by the unit price of that product. 

So we can define the problem, using a *Cobb-Douglas* utility function of 4 products, as follows:

$$max_{(x,y,z,w)}U = x^{\alpha}y^{\beta}z^{\gamma}w^{\delta}$$


Income Constraint: 

$$I = p_x*x + p_y*y +p_zz +p_ww$$

where $U$ = utility

$x$, $y$, $z$ and $w$ are four different goods or services

$I$ = income of the individual

$p_x$ = price of a unit of good/service $x$

$p_y$ = price of a unit of good/service $y$

$p_z$ = price of a unit of good/service $z$

$p_w$ = price of a unit of good/service $w$

We can define the problem in python using the *spicy* library of optimization. Because this library minimizes, we need to define the problem such that the utility function is minimize. We used the following utility function of a certain individual

$$min_{(x,y)}U = -(x^{2/3}y^{1/3}z^{2/7}w^{1/8})$$

constrained by:

$$- I = - p_xx - p_yy - p_zz - p_ww$$

And we need to define the variables $p_x, p_y$ $p_y$, $p_z$ and $I$. 

If we say that the budget an individual is equal to **1500€**. And we have four different products: 

Product $X$ is <a href = "https://www.amazon.com/-/es/Starbucks-C%C3%A1psulas-K-Cup-tostado-franc%C3%A9s/dp/B06XXGP24J/ref=sr_1_11?__mk_es_US=%C3%85M%C3%85%C5%BD%C3%95%C3%91&crid=265JGKOU1K90Q&keywords=starbucks+k+cups&qid=1638636238&sprefix=starbucks+%2Caps%2C295&sr=8-11" target = "_blank">*Starbuck's French Roast coffee capsules (60 pods)*</a>, which has a price of **54.99€** per unit.

The second product $Y$ is <a href = "https://www.amazon.com/-/es/Benefit-Cosmetics-Gimme-volumen-Marr%C3%B3n/dp/B01MTIXYKI/ref=sr_1_3?__mk_es_US=%C3%85M%C3%85%C5%BD%C3%95%C3%91&keywords=benefit&qid=1638636492&sr=8-3" target = "_blank">*Benefit Cosmetics Gimme - Brow Gel*</a>, which has a price of **22.99€** per unit. 

The third product $Z$ is an *All You Can Eat* Sushi Buffet menu, that has a price of **13.99€** per menu (per unit). 

And the fourth product $W$ a <a href = "https://www.zara.com/es/es/sudadera-b%C3%A1sica-capucha-p01660813.html?v1=153713350&v2=1882974" target = "_blank" >ZARA Sweatshirt</a> that has a price of **15.95€**

We have the following values: 

$p_x = 54.99$

$p_y = 22.99$

$p_z = 13.99$

$p_w = 15.95$

$I = 1500$. 

And the problem is equal to:

$$min_{(x,y)}U = -(x^{2/3}y^{1/3}z^{2/7}w^{1/8})$$

constrained by:

$$1500 - 54.99x - 22.99y - 13.99z - 15.95w= 0 $$

In [3]:
import numpy as np
from scipy.optimize import minimize

def objective(X):
    x, y,z,w= X
    return -(x**(2/3)*y**(1/3)*z**(2/7)*w**(1/8))

def gradient(X):
    x, y ,z,w= X
    return np.array([ -(2/3)*(x**(2/3 -1))*y**(1/3)*z**(2/7)*w**(1/8), -(1/3)*(x**(2/3))*(y**(1/3 -1))*z**(2/7)*w**(1/8), -(2/7)*(x**(2/3))*(y**(1/3))*(z**(2/7-1))*w**(1/8), -(1/8)*(x**(2/3))*(y**(1/3))*(z**(2/7))*w**(1/8 -1)])

def eq_cons(X):
    x, y,z,w = X
    return 1500 - 54.99*x -22.99*y - 15.95*z - 13.99*w

cons = ({'type': 'eq',
    'fun' : eq_cons})

sol = minimize(objective, x0=[1, 1,1,1], constraints=cons, jac=gradient, 
               method='SLSQP', options={'disp': True})
sol

Optimization terminated successfully.    (Exit mode 0)
            Current function value: -42.07886707116083
            Iterations: 15
            Function evaluations: 15
            Gradient evaluations: 15


     fun: -42.07886707116083
     jac: array([-2.17603492, -0.90978762, -0.6313396 , -0.55361763])
 message: 'Optimization terminated successfully.'
    nfev: 15
     nit: 15
    njev: 15
  status: 0
 success: True
       x: array([12.89160286, 15.4171025 , 19.04289456,  9.50088663])

To maximize the utility (satisfaction/happiness) of this individual, they need to buy $12.89$ units of product $X$, $15.41$ units of product $Y$, $19.04$ units of service $Z$ and $9.5$ units of product $W$. 

If we analyze these values, we can say that in this individual's case, the alternative with the highest total utility is product $Z$, which is the **All You Can Eat Sushi Menu**, which means that is the product that produces the higher satisfaction in this individual within their budget/income. It also means that product $Z$ is preferred by the individual 1.5 times more than product $X$, 1.2 times more than product $Y$ and 2.1 times more than product $W$. 

We can conclude that this individual gets more utility from the service product (eating outside of the house) than the self-care and food products (makeup, coffee and clothes). 